samples/JobsAgent/JobsAgentOperations.cpp

```diff
@@ -0,0 +1,693 @@
+                    util::String fileNameWithPath = GetFullPath(workingDirectory, files[i]["fileName"].GetString());
+                    IF_UNIT_TESTS(operation_status_ += fileNameWithPath + ';');
+
+                    std::ifstream src(fileNameWithPath, std::ios::binary);
+                    if (src.good()) {
+                        std::ofstream dst(fileNameWithPath + ".old", std::ios::binary);
+                        if (dst.fail()) {
+                            statusDetailsMap["errorCode"] = "ERR_FILE_COPY_FAILED";
+                            statusDetailsMap["errorMessage"] = "unable to backup file";
+                            statusDetailsMap["fileName"] = fileNameWithPath;
+                            return ResponseCode::FAILURE;
+                        }
+                        IF_NOT_UNIT_TESTS(dst << src.rdbuf());
+                    }
+                }
+            }
+
+            return ResponseCode::SUCCESS;
+        }
+
+        ResponseCode JobsAgent::RollbackFiles(util::Map<util::String, util::String> & statusDetailsMap,
+                                              util::String workingDirectory,
+                                              util::JsonValue & files) {
+            AWS_LOG_INFO(LOG_TAG_JOBS_AGENT, "RollbackFiles");
+            IF_UNIT_TESTS(operation_status_ += "RollbackFiles:");
+
+            ResponseCode rc = ResponseCode::SUCCESS;
+
+            statusDetailsMap["step"] = "rollback files";
+
+            for (unsigned int i = 0; i < files.Size(); i++) {
+                if (files[i].HasMember("fileName") && files[i]["fileName"].IsString()) {
+                    util::String fileNameWithPath = GetFullPath(workingDirectory, files[i]["fileName"].GetString());
+                    IF_UNIT_TESTS(operation_status_ += fileNameWithPath + ';');
+
+                    std::ifstream src(fileNameWithPath + ".old", std::ios::binary);
+                    if (src.good()) {
+                        std::ofstream dst(fileNameWithPath, std::ios::binary);
+                        if (dst.fail()) {
+                            statusDetailsMap["rollbackError"] = "not all files were successfully rolled back";
+                            rc = ResponseCode::FAILURE;
+                        } else {
+                            IF_NOT_UNIT_TESTS(dst << src.rdbuf());
+                        }
+                    }
+                }
+            }
+
+            return rc;
+        }
+
+        ResponseCode JobsAgent::DownloadFiles(util::Map<util::String, util::String> & statusDetailsMap,
+                                              util::String workingDirectory,
+                                              util::JsonValue & files) {
+            AWS_LOG_INFO(LOG_TAG_JOBS_AGENT, "DownloadFiles");
+            IF_UNIT_TESTS(operation_status_ += "DownloadFiles:");
+
+            statusDetailsMap["step"] = "download files";
+
+            CURLcode res;
+            CURL *curl = curl_easy_init();
```

Comment:
doesn't the cpp sdk already have an http client built in to it?

Reply:
Not the device SDK.